sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -129,6 +132,32 @@ class RelationalGroupedDataset protected[sql](
     (inputExpr: Expression) => exprToFunc(inputExpr)
   }
 
+  /**
+   * Returns a `KeyValueGroupedDataset` where the data is grouped by the grouping expressions
+   * of current `RelationalGroupedDataset`.
+   *
+   * @since 3.0.0
+   */
+  def keyAs[U : Encoder]: KeyValueGroupedDataset[U, T] = {
+    val keyEncoder = encoderFor[U]
+    val aliasedGrps = groupingExprs.map { g =>
+     g.transformDown {
+       case u: UnresolvedAttribute => df.resolve(u.name)
+     }
+    }.map(alias)
+    val additionalCols = aliasedGrps.filter(g => !df.logicalPlan.outputSet.contains(g.toAttribute))
+    val qe = Dataset.ofRows(
+      df.sparkSession,
+      Project(df.logicalPlan.output ++ additionalCols, df.logicalPlan)).queryExecution
```

Comment:
This seems inefficient. Can we make `KeyValueGroupedDataset.groupingAttributes` a `Seq[Expression]`?

Reply:
Will it also break break source compatibility?

Reply:
`groupingAttributes` is private in `KeyValueGroupedDataset`

Reply:
`KeyValueGroupedDataset` does not produce grouping attributes. These grouping attributes still come from the given `queryExecution`.

If we change `groupingAttributes` to `Seq[NamedExpression]`, we still need add this `Project` to produce these grouping attributes.

Reply:
Why do we need the grouping attributes? AFAIK it's used to specify the required distribution, which doesn't have to be attributes.

Reply:
For example, we use `groupingAttributes` to construct `MapGroups` in `flatMapGroups`.  These should be attributes so the `UnresolvedDeserializer` can be bound.

Reply:
We can make `MapGroupsExec` follow aggregate, to calculate grouping columns ahead and put it in the buffer row, then run key/value deserializer on buffer row.

But I admit that it's a big change, we can do it in the future.